### PR TITLE
iProc PCIe:  Add MSI-X Per-Vector Masking Support

### PR DESCRIPTION
**Pull request by Abhishek-brcm**
Brief on MSI-X Per-Vector Masking:
Per-vector masking(PVM) is managed through a Mask and Pending bit pair per MSI-X Table entry. An MSI-X vector is masked when its associated MSI-X Table entry Mask bit or the MSI-X Function Mask bit in the PCIe Configuration space is set. While a vector is masked, the function is prohibited from sending the associated message, and the function must set the associated Pending bit whenever the function would otherwise send the message. When host software unmasks a vector whose associated Pending bit is set, the function must schedule sending the associated message, and clear the Pending bit as soon as the message has been sent.

What we have done? :
Add support for interrupt driven MSI-X PVM feature for Viper.
Function mask bit update is tracked with snoop interrupt
and vector mask bit update is tracked with pcie pmon lite
address range access detection interrupt.
Both the interrupts are required to enable this feature.

Below commits achieve this:
drivers: pcie_ep: iproc: Add MSI-X PVM feature for Viper
dts: arm: viper: m7: Add PVM interrupts for pcie node

Some additional commits around earlier added reset interrupt support are also added:
dts: arm: viper: m7: Add reset interrupts for pcie node
drivers: pcie_ep: iproc: enable reset interrupts conditionally

And finally, couple of cleanup commits:
drivers: pcie_ep: iproc: shorten file names
drivers: pcie_ep: iproc: move msi(x) functions

### CODE REVIEW

**Review by dcpleung — CHANGES_REQUESTED** [review]
The `BCM` in kconfigs and filenames are to identify those to be targeting Broadcom devices. I suggest keeping them there.

**Comment by Abhishek-brcm** [comment]
> The `BCM` in kconfigs and filenames are to identify those to be targeting Broadcom devices. I suggest keeping them there.

The filenames/config names were just too long. "iProc" is sufficient to indicate it's Broadcom device. This matches with our Linux PCIe RC driver naming (see `drivers/pci/controller/pcie-iproc.c`). We have been submitting more drivers with similar naming convention (e.g. https://github.com/zephyrproject-rtos/zephyr/pull/28010/)

**Review by dcpleung — APPROVED** [review]

File: drivers/pcie/endpoint/pcie_ep_iproc_msi.c [hunk 1]
```diff
@@ -93,14 +91,140 @@ int iproc_pcie_generate_msix(const struct device *dev, const uint32_t msix_num)
...
+
+	/*
+	 * Read function mask bit/vector mask bit and update pending bit
+	 * with spin_lock - aim is not to allow interrupt context
+	 * to update PBA during this section
+	 * This will make sure of no races between mask bit read
+	 * and pending bit update.
+	 */
+
+	key = k_spin_lock(&ctx->pba_lock);
```
Comment (opsound) [inline drivers/pcie/endpoint/pcie_ep_iproc_msi.c:217]: A bit strange that this is the only place the driver is taking `pba_lock`. Typically they need to be taken in both thread and interrupt context to be safe on SMP and non-SMP.

Comment (Abhishek-brcm) [reply drivers/pcie/endpoint/pcie_ep_iproc_msi.c:217]: Yes right, I have considered only uniprocessor system as is our use case. I will look into making this SMP/multicore proof.

**Comment by carlescufi** [comment]
@Abhishek-brcm anyone else who can review this PR?

File: drivers/pcie/endpoint/pcie_ep_iproc_msi.c (hunk 1 above)
Comment (Abhishek-brcm) [reply drivers/pcie/endpoint/pcie_ep_iproc_msi.c:227]: @opsound, can you please have a look now?

Comment (opsound) [reply drivers/pcie/endpoint/pcie_ep_iproc_msi.c:227]: LGTM

**Review by opsound — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**